src/lib.rs

```diff
@@ -45,14 +45,22 @@ pub const MAX_DEPTH: usize = 32;
 /// Define the MAX supported thread name length. TODO: make this variable mutable.
 pub const MAX_THREAD_NAME: usize = 16;
 
+#[cfg(target_os = "linux")]
+mod addr_validate;
```

Comment:
I think we should also provide `addr_validate` for macOS, this is possible. The only problem is that macOS doesn't have `pipe2`, but we can simply simulate it.

Here is a possible piece of code:
```rust
#[inline]
#[cfg(target_os = "linux")]
unsafe fn create_pipe(fds: *mut libc::c_int) -> libc::c_int {
    libc::pipe2(fds, libc::O_CLOEXEC | libc::O_NONBLOCK)
}

#[cfg(target_os = "macos")]
unsafe fn create_pipe(fds: *mut libc::c_int) -> libc::c_int {
    let res = libc::pipe(fds);
    if res != 0 {
        return res;
    }
    let fds = fds as *mut [libc::c_int; 2];
    for n in 0..2 {
        let mut flags = libc::fcntl((*fds)[n], libc::F_GETFD);
        flags |= libc::O_CLOEXEC;
        let res = libc::fcntl((*fds)[n], libc::F_SETFD, flags);
        if res != 0 {
            return res;
        }
        let mut flags = libc::fcntl((*fds)[n], libc::F_GETFL);
        flags |= libc::O_NONBLOCK;
        let res = libc::fcntl((*fds)[n], libc::F_SETFL, flags);
        if res != 0 {
            return res;
        }
    }
    0
}
```

Reply:
I rework an implementation with `nix`. Could you test it on Mac OS (as I don't have a mac environment)?

Reply:
I ran `cargo test --features=flamegraph` on ARM64 macOS and got `test addr_validate::test::failed_validate has been running for over 60 seconds`. I guess the problem happens because we haven't set `O_NONBLOCK` for pipe on macOS at [src/addr_validate.rs#L21-L38](https://github.com/tikv/pprof-rs/pull/116/files#diff-9a313809f180a1657b0b852dd388e0aef3f17884806d935429622884a7080fd5R21-R38).

Reply:
@mornyx PTAL

Reply:
`FdFlag::O_NONBLOCK` does not exist, I have made a suggestion below.